Test development guidelines final (#25)
mysugr/sweetest: GUIDELINES.md

### Add configuration in the steps class

In order to know how we set up the test we should first quickly get a grasp of the system we want to test. Here is an outline of the example with a typical Android architecture:

```
┏━━━━━━━━━━━━━━━┓
┃ LoginActivity ┃  ⟵ Android view
┗━━━━━━━━━━━━━━━┛

━━━ ᐁ UNDER TEST ᐁ ━━━

┏━━━━━━━━━━━━━━━━┓
┃ LoginViewModel ┃  ⟵ presentation
┗━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━┓ 
┃ AuthManager ┃  ⟵ domain
┗━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━━┓
┃ BackendGateway* ┃ ┃ SessionStore* ┃  ⟵ data access
┗━━━━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━━┛
* = mocked
```

In order to achieve that you have to add a configuration that reflects the wanted setup in the steps file:

```kotlin
override fun configure() = super.configure()
    .requireReal<LoginViewModel>()
    .requireReal<AuthManager>()
```

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Author: `any` just means that the dependency has the chance to become both a mock and a real instance. but it can be determined at test run time.

so no, it works also if you e.g. use `mockOnly` or `realOnly`, then you don't need to use `requireMock` or `requireReal` in steps/test classes.

does this answer your question? i'm not 100% sure i understood your question correctly.

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Reviewer: So when I don't have `any` in the configuration, but `requireReal`, it runs?  I didn't get to reference yet, but a table there with what works and what not would be nice :)

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Author: i again don't understand your question 🤣 what do you mean by "don't have `any`"? when you have `any` in the module config it's the least constraining you can have. so `requireReal` down in the steps/class file would work. just if you set `mockOnly` in the module config and `requireReal` it's a conflicting configuration and then it fails.

i don't know why we should add a comparison table here. first, tha place you tagged is probably the wrong one her, and second, dependency configuration on a per-module basis will be deprecated. but maybe i didn't understand fully what you mean (still 😁), if you want we can have a call 😉

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Reviewer: I just remember my sweetest journey and that it is easy to forget adding the dependency in the moduleConfiguration. So a sentence like: "Also make sure the dependency you add here with `requireReal` has to be defined in the module config, see above." 

A table could be added in Reference I meant. Here it doesn't make sense of course.

> second, dependency configuration on a per-module basis will be deprecated.
That was actually my question, whether this will be improved. Nice to hear that 👍

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Author: > I just remember my sweetest journey and that it is easy to forget adding the dependency in the moduleConfiguration. So a sentence like: "Also make sure the dependency you add here with `requireReal` has to be defined in the module config, see above."

makes total sense 👍 

> A table could be added in Reference I meant. Here it doesn't make sense of course.

you mean something like this?

| Configuration | Module testing configuration | allowed? |
|:---|:--|:--|
| `requireMock` | `any` | ✅ (`any` doesn't constraint mode) |
| `requireMock` | `realOnly` | ❌ (`realOnly` doesn't allow mocks) |
| `requireReal` | `realOnly` | ✅ (`realOnly` allows `requireReal`) |
| `requireReal` | `mockOnly` | ❌ (`mockOnly` doesn't allow `requireReal`)|
| `requireReal` | `any` | ✅ (`any` doesn't constraint mode) |

i think that's redundant somehow...

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Reviewer: Haha, yes exactly. Nice that you did it and then say it is redundant. I would have order the columns differently, module testing config first and then Steps Configuration. There would be even more situations (e.g. realOnly and no config in steps etc). Instead of allowed, the last column would be: It is mock or real or not allowed. I would like a table like that as reference a lot, but just my opinion

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Author: ok, for time's sake i moved it to my notes so it's respected in the next update, which will anyways happen soon because of the mockk-update. reason: i need to wrap my head around the upcoming deprecation of the dependency configuration in the module testing configuration. i hope that's ok for you!

### Add access to the dependencies in the steps class

> When we look at the test case we've written above we can see that we want to put something into the top of the system under test (the view model) and see what comes out at the bottom (backend gateway). So let's add members to the class that enable us to interact with them:

Author: it depends... if you just want to see if the view model ends up with the correct status then yes. i think that's something that could be added to the example. maybe we can leave that to future improvements...

it depends... if you need to control the environment there is two possibilities: the `AuthManager` or the `BackendGateway` (and eventually also the `SessionStore`). if you want to do the test in an integrated fashion you need to go down to the `BackendGateway`. but anyways it's abstracted behind a `BackendFakeSteps` that hides the implementation details.

is that sufficient for you?

> When we look at the test case we've written above we can see that we want to put something into the top of the system under test (the view model) and see what comes out at the bottom (backend gateway). So let's add members to the class that enable us to interact with them:

Reviewer: Yeah, as said out-of-scope. And obviously it always depends. :)

1. We can create a steps which resembles a fake version of the backend gateway that is capable of more sophisticated tasks like having fake users and acting upon that data
3. We can create a test and steps class pair which focuses on unit-testing the `AuthManager` that uses the fake backend steps class
4. We can re-wire the previous integration test to use the same fake backend steps class

So in one word: we're going to show sweetest's strengths by reusing test code.

#### The fake backend

The concept of steps classes is that they include everything needed to add a piece to the test system. In this example that will be:

* Configuration
* A fake implementation of a production interface
* Code to set up the fake
* Code to assert calls on the fake

Therefore this is how our `BackendFakeSteps` class looks like:

```kotlin

> class BackendFakeSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

Author: i figured that everyone is approaching fakes a little different therefore i thought alignment wouldn't hurt. also in some articles i read that stubbing (a la mockito) should not be overused and fake implementations should be used when it serves the value of the test. at least it broadens the perspective on the topic. do you have a strong opinion on whether to leave it in the guidelines?

> class BackendFakeSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

Reviewer: No strong opinion. Just thought it makes it a bit easier to follow, but up to you

        val users = mutableListOf<BackendFakeUser>()

        override fun checkEmail(email: String): Boolean = users.find { it.email == email } != null

        override fun login(email: String, password: String): AuthToken {
            val user = users.find { it.email == email && it.password == password }
                ?: throw UsernameOrPasswordWrongException()
            return user.authToken
        }

        override fun register(email: String, password: String): AuthToken {
            TODO()
        }

        override fun getUserData(authToken: AuthToken): User {
            TODO()
        }
    }
}
```

Maybe you noticed that the naming `BackendFakeSteps` doesn't refer to the `BackendGateway` whose behavior it tries to mimic. This is intentional: this ties it more to the abstract concept of a "backend" than to how its access is technically implemented. So if we decide to come up with a different way of accessing the backend in the future, the steps classes' API and the tests using it can remain unchanged. Therefore also the steps classes' API doesn't reveal much about the technicalities.

> If you're wondering how the `BackendFakeUser` looks like, here it is:

Author: i think it depends: if you are trying to model unique entities like "users" in this case it makes sense to mirror that by having "const"-like declaration. for generating other test data i'd totally agree. in the context of the introduction it shows how one can think of reuse also with respect to test _data_.

#### Create the steps class for the unit test

So let's create a steps class responsible for just unit-testing `AuthManager`:

```kotlin
class AuthManagerSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

    val backend by steps<BackendFakeSteps>()

    private val instance by dependency<AuthManager>()
    private val sessionStore by dependency<SessionStore>()

    override fun configure() = super.configure()
        .requireReal<AuthManager>()
}
```

As we anticipate that the fake backend and the `SessionStore` will be needed afterwards (they are the two direct dependencies of `AuthManager` which are consumed by its constructor).

There is only one `requireReal` call (`requireReal<AuthManager>()`), which clearly indicates that this steps class is indeed about unit-testing just the `AuthManager`.

> But why is `backend` just called `backend` and not `backendGateway`? And why is it public? Shouldn't a steps class be a full abstraction of what's going on under the hood or encapsulate everything needed for the test class using it? Generally yes. Usually we should avoid offering internals ([Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)). Access to internals should be done through public accessors in order to avoid tying code to details which are prone to change. But anyway, the key here is "prone to change": if you group and define systems around abstract business concepts these are very unlikely to change (contrary to technical implementation). This makes the fake backend a better candidate for being used and shared among various tests. So in this case it is reasonable to call the member after the abstract concept of a "backend" _and_ offering it via the steps classes' API for direct use in test classes.

Author: i agree that this touches quite a lot. it is very important for me to get the whole abstraction topic across, as it's the whole point of sweetest, right? this was the original intention behind this sentence. could you offer me a better way of achieving this?

> But why is `backend` just called `backend` and not `backendGateway`? And why is it public? Shouldn't a steps class be a full abstraction of what's going on under the hood or encapsulate everything needed for the test class using it? Generally yes. Usually we should avoid offering internals ([Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)). Access to internals should be done through public accessors in order to avoid tying code to details which are prone to change. But anyway, the key here is "prone to change": if you group and define systems around abstract business concepts these are very unlikely to change (contrary to technical implementation). This makes the fake backend a better candidate for being used and shared among various tests. So in this case it is reasonable to call the member after the abstract concept of a "backend" _and_ offering it via the steps classes' API for direct use in test classes.

Reviewer: > it is very important for me to get the whole abstraction topic across, as it's the whole point of sweetest, right?

Fair enough. I would kept these ideas still more marked as tip. As especially for the introduction it might be a bit overwhelming. My suggestion was the one sentence in my comment, but up to you.

> But why is `backend` just called `backend` and not `backendGateway`? And why is it public? Shouldn't a steps class be a full abstraction of what's going on under the hood or encapsulate everything needed for the test class using it? Generally yes. Usually we should avoid offering internals ([Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)). Access to internals should be done through public accessors in order to avoid tying code to details which are prone to change. But anyway, the key here is "prone to change": if you group and define systems around abstract business concepts these are very unlikely to change (contrary to technical implementation). This makes the fake backend a better candidate for being used and shared among various tests. So in this case it is reasonable to call the member after the abstract concept of a "backend" _and_ offering it via the steps classes' API for direct use in test classes.

Author: hm, i think i found a way to accomodate your feedback here: i moved the monster sentence to the reference section (namely it fits perfectly in the "Can there be too much abstraction?" chapter).

        backend.givenExistingUser(USER_A)
        whenPassingCredentials(USER_A.email, USER_A.password)
        thenSessionWasStarted()
    }
}
```

As you can see, `backend` is directly used and this will allow for a lot of code reuse. So let's see how that simplifies the integration test we've created in the beginning:

#### Use the fake backend in the integration test

Let's alter the steps class for the login test so it uses the fake backend:

```kotlin
class LoginSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

    val backend by steps<BackendFakeSteps>() // <-- added

    private val viewModel by dependency<LoginViewModel>()
//  private val backendGateway by dependency<BackendGateway>() <-- removed

    override fun configure() = super.configure()
        .requireReal<LoginViewModel>()
//      .requireReal<AuthManager>() <-- instead of configuring it here, we just...
        .requireSteps<AuthManagerSteps>() // <-- ...include the steps class which does the configuration

    fun whenLoggingIn(email: String, password: String) {
        viewModel.loginOrRegister(email, password)
    }

// Not needed anymore:

> //  fun givenExistingUser(email: String, password: String, authToken: AuthToken) {

Author: i have a really hard time getting rid of the commented-out code. it's analogous to "a picture tells more than thousand words". some parts of the guidelines is still hard to understand, so i want that to really get across. just asking: which harm is it really doing? i'd like to understand...

```kotlin
class LoginTest : BaseJUnitTest(appModuleTestingConfiguration) {

    private val sut by steps<LoginSteps>()

    @Test
    fun `Logging in checks email at backend`() {
        sut {
            backend.givenExistingUser(USER_A)
            whenLoggingIn(USER_A.email, USER_A.password)
            backend.thenEmailWasChecked(USER_A.email)
        }
    }

// Before:

//  @Test
//  fun `Logging in checks email at backend`() {
//      sut {
//          givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
//          whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
//          thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
//      }
//  }

// Not needed anymore:

//  companion object {
//      const val EXISTING_EMAIL = "existing@test.com"
//      const val EXISTING_PASSWORD = "supersecure1"
//      const val EXISTING_AUTH_TOKEN = "auth_token"
//  }
    
}
```

### Summing up

> In this journey it becomes apparent that having integration tests can be simplified a lot when there is already a steps class that tests a smaller unit (`AuthManagerSteps`): it basically requires just using (`... by steps<AuthManagerSteps>()`) or including (`requireSteps<AuthManagerSteps>`) the respective steps class and as a result instead of using a mock there is suddenly a real instance of the class (`AuthManager`) in the test system with just very low effort.

Reviewer: I could imagine that summing up could be a bit shorter and not explaining the details again. We showed: 3-5 bullet points. (Bullet points are beautiful and give a nice overview)

Author: i tried to boil it down in some bullet points, hope you like it 🙂

The fake backend example also shows how code reuse can work very easily. In this example we were able to reduce the test code by more or less a half.

> All this of course comes with a cost: steps classes have to be set up and very well thought through and preferably be modelled around abstract business models (`BackendFakeSteps`, `LoginSteps`), but that's not always possible or even reasonable (e.g. the `AuthManagerSteps` is a direct reference to a class, which is OK as it serves as an abstraction layer in two test classes).

Reviewer: This might also be in an extra section. Instead of adding this to summing up, it could be in something like `Things to look out` and laid out in bullet points

Author: after implementing your suggestion i think this is somehow obsolete?

  * so test code, especially its technical implementation (steps classes), can be reused
* **Simplify dependency tree creation** by using configuration and automatic dependency resolution
* **Test setup is simplified** so **integration tests become default**
  * which **reduces the use of mocks** and
  * leads to **more realistic tests**

## Introduction

This introduction guides you through the setup of a typical sweetest test. It's goal is to be as comprehensive and reduced as possible for people not yet familiar with sweetest. We recommend reading through the [reference section](#reference) too in order to deepen your understanding of sweetest and generally using it any time you need during development.

### Use the live templates!

If you use [IntelliJ IDEA](https://www.jetbrains.com/idea) (respectively [Android Studio](https://developer.android.com/studio)) as IDE you should use these [live templates](https://www.jetbrains.com/help/idea/using-live-templates.html) to speed up development with sweetest:

<img src="readme/live-remplates-ssteps.gif">

> To make use of them please [download these exported IntelliJ settings](raw/master/tools/live-templates.jar) and import them into your IDE with `File / Import settings`. After the import you can see them in the settings at `Editor / Live Templates`:

Author: no, i think you're right, i'll change it!

#### Add the `when` function

Functions starting with `when` should take care of the "acting" part in "arrange", "act", "assert" and would hence interact directly with the production system; so let's wire the test to the production code:

```kotlin
fun whenLoggingIn(email: String, password: String) {
    viewModel.loginOrRegister(email, password)
}
```

#### Add the `then` function

Functions starting with `then` should solely do assertions:

```kotlin
fun thenEmailWasCheckedAtBackend(email: String) {
    verify(backendGateway).checkEmail(email)
}
```

### Current state

```kotlin
class LoginTest : BaseJUnitTest(appModuleTestingConfiguration) {

    private val sut by steps<LoginSteps>()

    @Test
    fun `Logging in checks email at backend`() = sut {
        givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
        whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
        thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
    }

    companion object {
        const val EXISTING_EMAIL = "existing@test.com"
        const val EXISTING_PASSWORD = "supersecure1"

>         const val EXISTING_AUTH_TOKEN = "auth_token"

Author: i'm a little split here. as you said it's a matter of taste. but in this case it also has the advantage that it shows better how much code is reused as the declarations you marked vanish.

class BackendFakeSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

    // Creates a fake instance and wrap it with a Mockito spy (for the sake of being able to use `verify`)
    private val instance = spy(FakeBackendGateway())

    // Provide an instance of `BackendGateway` to sweetest's dependency management
    override fun configure() = super.configure()
        .offerMockRequired<BackendGateway> { instance }

    // This adds a user to the fake backend
    fun givenExistingUser(backendFakeUser: BackendFakeUser) {
        instance.users += backendFakeUser
    }

    // This verifies an expected call to the fake
    fun thenEmailWasChecked(email: String) {
        verify(instance).checkEmail(email)
    }

    // Same here
    fun thenLoginWasAttempted(email: String, password: String) {
        verify(instance).login(email, password)
    }

    // The fake is private as a steps classes' aim is to abstract the technical implementation of test code
    private class FakeBackendGateway : BackendGateway {

        // We leave the internals public for the outside class for simplicity's sake
        val users = mutableListOf<BackendFakeUser>()

>         override fun checkEmail(email: String): Boolean = users.find { it.email == email } != null

Reviewer 2: `!= null` is not necessary with:
```suggestion
        override fun checkEmail(email: String): Boolean = users.any { it.email == email }
```

  * leads to **more realistic tests**

## Introduction

This introduction guides you through the setup of a typical sweetest test. Its goal is to be as comprehensive and focused as possible for people not yet familiar with sweetest. We recommend reading through the [reference section](#reference) too in order to deepen your understanding of sweetest and generally using it any time you need during development.

### Use the live templates!

If you use [IntelliJ IDEA](https://www.jetbrains.com/idea) (respectively [Android Studio](https://developer.android.com/studio)) as IDE you should use these [live templates](https://www.jetbrains.com/help/idea/using-live-templates.html) to speed up development with sweetest:

<img src="readme/live-remplates-ssteps.gif">

To make use of them please [download these exported IntelliJ settings](raw/master/tools/live-templates.jar) and import them into your IDE with `File / Import settings`. After the import you can see them in the settings at `Editor / Live Templates`:

<img src="readme/live-templates-settings.png">

You can use the templates by beginning to type the abbreviations as shown in the image above and hitting enter.

> ### Add a module configuration

Author: * i added a link to the reference
* and there i added a broader description of the configuration
* the whole file example is a few lines below

is that enough you think?

Given you have a module `app` you should create a file `AppModuleTestingConfiguration.kt` in the root package of your module, e.g. `com.example.app`.

```kotlin

> val appModuleTestingConfiguration = moduleTestingConfiguration { ... }

Author: i think i don't understand. can you give me an example?

sweetest puts dependencies together for you by configuration. If you have a complex system under test you don't have to create the dependencies manually. In case a type is required by your test, the internal dependency management of sweetest examines the constructors of the dependencies and tries to satisfy all parameters. This is a recursive process that goes on until all dependencies are created.

That's how you declare dependencies in sweetest:

```kotlin
val appModuleTestingConfiguration = moduleTestingConfiguration {
    dependency any of<LoginViewModel>()
    dependency any of<AuthManager>()
    dependency any of<BackendGateway>()
    dependency any of<SessionStore>()
}
```

Put all dependencies in there as you need them to be auto-created by sweetest. E.g. if `LoginViewModel` requires `AuthManager` in its constructor you should add `AuthManager` to the dependency configuration, and so on...

### Create a steps class

The steps class will contain the _technical implementation_ of your test:

```kotlin
package com.example.app.view

class LoginSteps(testContext: TestContext)
    : BaseSteps(testContext, appModuleTestingConfiguration)
```

> You should put that class in the same package as the class with the highest abstraction level (`LoginViewModel`), in this case `com.example.app.view`.

Author: no, we decided against that. so test (and also steps) classes should be at the same package as the production type (even though they are in a different module or source st)

**Tip:** you can write all your test cases like that in the test class first and then let the IDE create the missing functions in the respective steps class (e.g. in IntelliJ: Option + Enter, select "Create member function" and the target class, in this case the steps class - and done!). That way you can flesh out the tests first and make the sources compilable very fast.

### Add configuration in the steps class

In order to know how to set up the test system we should first quickly get a grasp of the system we want to test. Here is an outline of the example with a typical MVVM architecture:

```
┏━━━━━━━━━━━━━━━┓
┃ LoginActivity ┃  ⟵ view
┗━━━━━━━━━━━━━━━┛

━━━ ᐁ UNDER TEST ᐁ ━━━

┏━━━━━━━━━━━━━━━━┓
┃ LoginViewModel ┃  ⟵ presentation
┗━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━┓ 
┃ AuthManager ┃  ⟵ domain
┗━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━━┓
┃ BackendGateway* ┃ ┃ SessionStore* ┃  ⟵ data access
┗━━━━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━━┛
* = mocked
```

> In order to achieve that you have to add a configuration that reflects the wanted setup in the steps file:

Reviewer 3: ```suggestion
To achieve this test setup you have to add a configuration that reflects the wanted setup in the steps class:
```

```kotlin
override fun configure() = super.configure()
    .requireReal<LoginViewModel>()
    .requireReal<AuthManager>()
```

This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks if the dependency isn't configured otherwise. So ultimately that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

### Add access to the dependencies in the steps class

> When we look at the test case we've written above we can see that we want to put something into the top of the system under test (the view model) and see what comes out at the bottom (backend gateway). So let's add members to the class that enable us to interact with them:

Reviewer 3: ```suggestion
When we look at the test case we've written above we can see that we want to start off by putting something into the system under test (the view model) and finally check the returned result in the end (backend gateway). Therefore, let's add members to the class that enable us to interact with them:
```

      - [Summing up naming](#summing-up-naming-2)
- [Links](#links)

## Goals

* Put a **layer of abstraction** on the system under test (SuT): we call these abstractions _steps_ (as derived from Cucumber)
  * so if the system under test changes, most parts of the test system don't need to change
  * so the test just tells _what_ is tested, not _how_ (tests become more business-centric; whenever possible, all _technical implementation_ is in the steps)
  * so test code, especially its technical implementation (steps classes), can be reused
* **Simplify dependency tree creation** by using configuration and automatic dependency resolution
* **Test setup is simplified** so **integration tests become default**
  * which **reduces the use of mocks** and
  * leads to **more realistic tests**

## Introduction

This introduction guides you through the setup of a typical sweetest test. Its goal is to be as comprehensive and focused as possible for people not yet familiar with sweetest. We recommend reading through the [reference section](#reference) too in order to deepen your understanding of sweetest and generally using it any time you need during development.

> ### Use the live templates!

Author: hm, but then there would be just one tip? 🤔

> ### Use the live templates!

Reviewer 3: donno maybe the templates or some common mistakes/problems/solutions.. i thought I've read it somewhere in the discussion here elsewhere :D

**Tip:** you can write all your test cases like that in the test class first and then let the IDE create the missing functions in the respective steps class (e.g. in IntelliJ: Option + Enter, select "Create member function" and the target class, in this case the steps class - and done!). That way you can flesh out the tests first and make the sources compilable very fast.

### Add configuration in the steps class

In order to know how to set up the test system we should first quickly get a grasp of the system we want to test. Here is an outline of the example with a typical MVVM architecture:

```
┏━━━━━━━━━━━━━━━┓
┃ LoginActivity ┃  ⟵ view
┗━━━━━━━━━━━━━━━┛

━━━ ᐁ UNDER TEST ᐁ ━━━

┏━━━━━━━━━━━━━━━━┓
┃ LoginViewModel ┃  ⟵ presentation
┗━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━┓ 
┃ AuthManager ┃  ⟵ domain
┗━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━━┓
┃ BackendGateway* ┃ ┃ SessionStore* ┃  ⟵ data access
┗━━━━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━━┛
* = mocked
```

To achieve this test setup you have to add a configuration that reflects the wanted setup in the steps class:

```kotlin

> override fun configure() = super.configure()

Author: the second bullet point is answered in the sentence after. is that enough to clarify your question?

    .requireReal<LoginViewModel>()
    .requireReal<AuthManager>()
```

> This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks if the dependency isn't configured otherwise. So ultimately that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.

Author: i think as for now it's just mockito mocks i'd like to leave it like that. when we change the implementation we should state in the beginning of the guidelines that we support more than just mockito. but until then removing "mockito" could be even more confusing...

In most cases `given` will define the behavior of the system before we really start interacting with the production system (the "acting" part) and thus controls the behavior of the mocks or fakes we have in our test system:

```kotlin
fun givenExistingUser(email: String, password: String, authToken: AuthToken) {
    `when`(backendGateway.checkEmail(email)).thenReturn(true)
    `when`(backendGateway.login(email, password)).thenReturn(authToken)
}
```

In this case the state of a user being existent is achieved by setting up the backend gateway so it responds with specific answers.

#### Add the `when` function

Functions starting with `when` are taking care of the "acting" part in "arrange", "act", "assert" and would hence interact directly with the production system; so let's wire the test to the production code:

```kotlin
fun whenLoggingIn(email: String, password: String) {
    viewModel.loginOrRegister(email, password)
}
```

#### Add the `then` function

Functions starting with `then` are solely concerned with assertions:

```kotlin
fun thenEmailWasCheckedAtBackend(email: String) {
    verify(backendGateway).checkEmail(email)
}
```

> ### Current state

Author: hm, or maybe `Current state of the example`?

        const val EXISTING_AUTH_TOKEN = "auth_token"
    }
}

class LoginSteps(testContext: TestContext) :
    BaseSteps(testContext, appModuleTestingConfiguration) {

    private val viewModel by dependency<LoginViewModel>()
    private val backendGateway by dependency<BackendGateway>()

    override fun configure() = super.configure()
        .requireReal<LoginViewModel>()
        .requireReal<AuthManager>()

    fun givenExistingUser(email: String, password: String, authToken: AuthToken) {
        `when`(backendGateway.checkEmail(email)).thenReturn(true)
        `when`(backendGateway.login(email, password)).thenReturn(authToken)
    }

    fun whenLoggingIn(email: String, password: String) {
        viewModel.loginOrRegister(email, password)
    }

    fun thenEmailWasCheckedAtBackend(email: String) {
        verify(backendGateway).checkEmail(email)
    }
}
```

### Improving structure by reusing test code

What if we want to create a simple unit test for `AuthManager`?

1. We can create a steps which resembles a fake version of the backend gateway that is capable of handling tasks like having fake users and acting upon that data.

> 3. We can create a test and steps class pair which focuses on unit-testing the `AuthManager` that uses the fake backend.

Author: i use 1. everywhere, the rest is done by the formatter 👍

4. We can re-wire the previous integration test to use the same fake backend steps class.

So in one word: we're going to show sweetest's strengths by reusing test code and working with dependencies distributed among multiple steps classes.

#### The fake backend

> The concept of steps classes is that they encapsulate everything needed to add a piece to the test system. In this example that will be:

Author: hm, or maybe that?

```
The aim of a steps class is to encapsulate test behavior which can subsequently be plugged into the test system wherever needed.

For the fake backend in this example that includes:
```

* Offering a fake implementation of a production interface
* Configuration (providing the fake to dependency management)
* Code to set up the fake
* Code to verify calls on the fake

> Therefore this is how our `BackendFakeSteps` class looks like:

Author: i thought about this and also there is a discussion in #24 and i decided for the better sorting (so it's sorted together with other types starting with `Backend`).

        val users = mutableListOf<BackendFakeUser>()

        override fun checkEmail(email: String): Boolean = users.any { it.email == email }

        override fun login(email: String, password: String): AuthToken {
            val user = users.find { it.email == email && it.password == password }
                ?: throw UsernameOrPasswordWrongException()
            return user.authToken
        }

        override fun register(email: String, password: String): AuthToken {
            TODO()
        }

        override fun getUserData(authToken: AuthToken): User {
            TODO()
        }
    }
}
```

Maybe you noticed that the name `BackendFakeSteps` doesn't reference the `BackendGateway` directly, whose behavior it implements. This is intentional: this ties it more to the abstract concept of a "backend" than to how its access is technically implemented. So if we decide to come up with a different way of accessing the backend in the future, the steps classes' API and the tests using it can remain unchanged. That is the reason why...

* the steps classes' API doesn't reveal much about the technicalities
* we keep the fake implementation private

> * the naming in the API is evolving around business terms as much as possible

Author: evolving --> revolving, my bad...

    val email: String,
    val password: String,
    val authToken: AuthToken = UUID.randomUUID().toString()
) {
    companion object {
        val USER_A = BackendFakeUser("user.a@test.com", "supersecure_a")
        val USER_B = BackendFakeUser("user.b@test.com", "supersecure_b")
    }
}
```

It also contains predefined data `USER_A` and `USER_B` so it can be reused by multiple tests without the need to define constants individually each time. Maybe you noticed that this class is outside the steps class; that is intentional too: its use reaches beyond the steps class as it can be handy for various test scenarios, no matter you decide to use the fake backend or not.

#### Create the steps class for the unit test

So let's create a steps class responsible for just unit-testing `AuthManager`:

```kotlin
class AuthManagerSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {

    val backend by steps<BackendFakeSteps>()

    private val instance by dependency<AuthManager>()
    private val sessionStore by dependency<SessionStore>()

    override fun configure() = super.configure()
        .requireReal<AuthManager>()
}
```

> The `AuthManager` type is of corse in the game, but also `SessionStore` and `BackendGateway` have to be considered, as they are two direct dependencies of `AuthManager`; so somehow that has to be reflected in the steps class:

Author: i used part of what you said, basically moved it up before the code block, better that way?

* `AuthManager` is configured to be  a real instance (`requireReal<AuthManager>()`) as it's obviously the class under test.
* Access to `AuthManager` is added (`val instance by dependency<AuthManager>()`) as there needs to be interaction with the instance.
* The `BackendGateway` is implemented as a fake by adding `val backend by steps<BackendFakeSteps>()`.

> * As writing the test you might figure out that also access to `SessionStore` is useful (`val sessionStore by dependency<SessionStore>()`): even though it's just a mock (as it's not configured otherwise) you still can perform verifications with it.

Author: hm, better that way? 

`As writing the test we might figure out that also access to `SessionStore` is useful (`val sessionStore by dependency<SessionStore>()`): in the example it's used to perform verifications on it.`

> * As writing the test you might figure out that also access to `SessionStore` is useful (`val sessionStore by dependency<SessionStore>()`): even though it's just a mock (as it's not configured otherwise) you still can perform verifications with it.

Author: i wanted to keep it simple for the sake of the example. but i added:

```
Also `SessionStore` is a perfect candidate for being faked in `SessionFakeSteps` or similar, but for the sake of simplicity we only concentrate on faking the backend.
```

//  }

}
```

As you can see, all the details regarding how `AuthManager` is being tested can go to the `AuthManagerSteps`, including configuration. And the integration test is still an integration test as `AuthManagerSteps` configures the `AuthManager` to be a real instance.

Additionally, we add access to `backend` and again offer it as a public member so it can be used in the test. Also we use `BackendFakeUser.USER_A` in order to not only reuse test code and configuration but also test _data_! Here's the result:

```kotlin
class LoginTest : BaseJUnitTest(appModuleTestingConfiguration) {

    private val sut by steps<LoginSteps>()

    @Test
    fun `Logging in checks email at backend`() = sut {
        backend.givenExistingUser(USER_A)
        whenLoggingIn(USER_A.email, USER_A.password)
        backend.thenEmailWasChecked(USER_A.email)
    }

// Before:

//  @Test
//  fun `Logging in checks email at backend`() = sut {
//      givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
//      whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
//      thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
//  }

> // Not needed anymore:

Reviewer 3: kinda the same topic as above.. I would maybe focus on the changes in the test (since using the fake user is already mentioned in the text).

I don't see value in advertising the code saving in a guideline since I think everyone uses sweetest anyway/already..

but maybe it makes sense to have this comparison in a split view (if that is somehow possible) to show the difference/progress and what enden up where (than i makes sense to me to show the whole file, as a snippet I would probably focus on the test)

Author: hm, at mySugr yes, but this is an open-source project 😁
let's continue on the other comment...

Reviewer 3: ah ok. wasn't aware of that :)

### Steps

Steps classes are means of **organizing and abstracting test code**. The easiest way to implement steps is to extract all technical implementation of tests into steps classes, leaving just business-facing function calls to the steps classes in the test class. This allows the test to look like this:

```kotlin
@Test
fun `Logging in checks email at backend`() = sut {
    givenExistingUser(USER_A)
    whenLoggingIn(USER_A.email, USER_A.password)
    thenEmailWasCheckedAtBackend(USER_A.email)
}
```

The steps class is then only concerned with the technical implementation.

The second aim of steps classes is to abstract not only the test code but also **configuration**. In order for the test class mostly being concerned with business-facing function calling, also the test setup should go to steps classes.

An instance of a steps class can only exist once during the whole test. So if the same type of steps class is requested multiple times always the same instance is returned.

Steps classes are initialized during a specific phase in the initialization of the framework and of course purged after each test function run to avoid side effects.

> By the way: the name "steps class" is taken from [Cucumber](https://cucumber.io/), a [behavior-driven testing](https://en.wikipedia.org/wiki/Behavior-driven_development) tool. Also there workflows are broken down into simple "steps" and grouped into steps classes. sweetest was in fact designed with interoperability with Cucumber in mind and takes the concepts further by allowing for interdependent steps classes and dependency management, where dependencies can be consumed across steps classes.

Author: i added a link from the introduction. but there is already a hint in the beginning of the intro to also read the reference.

#### Using steps classes

To create a steps class you have to derive from the `BaseSteps` class:

```kotlin
class LoginSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration)
```

You have to reference the module testing configuration (in this case `appModuleTestingConfiguration`) of the module the component under test lies in.

To consume a steps class you have to use the `steps` function inside a test or steps class:

```kotlin
val sut by steps<LoginSteps>()
```

By using the variable you have access to the members of the steps class. But if you don't need any access to the members and just want the steps class to load you can consider using `requireSteps` in the configuration:

```kotlin
override fun configure() = super.configure()
    .requireSteps<LoginSteps>()
```

### Dependencies

> sweetest is tailored for systems where [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is used. As you most likely have no DI during unit testing sweetest makes good for that by offering its own simple way of doing it.

Author: hm, i think that the abbreviation is quite common, also it's mentioned in the sentence right before...

> sweetest is tailored for systems where [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is used. As you most likely have no DI during unit testing sweetest makes good for that by offering its own simple way of doing it.

Reviewer 3: as far as I can see the abbr. is only used once and in the sentence before it is written as a whole. it probably is quite common but also in my opinion abbreviations should be avoided in documents or at least introduced.

since it is only used once I would just use the whole words I guess also because it adds to the reading flow. imo.

In sweetest all dependencies are treated as singletons, so there will be only one instance of a certain type. So it can only be used in these cases. In all other cases you have to fall back to managing object creation on your own. The automated way should cater for the very most cases, though.

#### Dependency modes

> For each dependency type there are two possibilities: either it's configured to be mock or real. Unless you tell sweetest otherwise, all dependencies are set to the "mock" mode by default.

Reviewer 2: ```suggestion
For each dependency type there are two possibilities: either it's configured to be mocked or real. Unless you tell sweetest otherwise, all dependencies are set to the "mock" mode by default.
```

##### Mock

If a dependency is set to the "mock" mode, a Mockito mock is created for the dependency.

* The mock is created lazily on demand

> * The same instance is cached when the same type is needed elsewhere

Reviewer 2: ```suggestion
* The instance is cached and reused when the same type is needed elsewhere
```

* The cache is cleared after each test function run

##### Real

What does it mean if you define a type as "real"?

Instance creation is different from "mock" here: the constructor of the _real_ class will be called. In case there are parameters they will be satisfied with arguments by sweetest automatically. All the arguments will be handled as dependencies exactly the same way in a recursive manner until the dependency graph is built up.

> Of course dependencies of dependencies can have different modes. E.g. the `LoginViewModel` can have mode "real", but the underlying dependencies can have mode "mock".

Reviewer 2: I find this sentence a bit confusing. For me it sounds like that I can have e.g. a real `AuthManager` but if it's a dependency of a dependency it can be also a mock at the same time.

Author: better now?

Reviewer 2: yes, better 👍

##### Configuring and requiring modes

> You should configure dependencies mostly in steps classes (but in very rare cases you might also need to do it in test classes). You can do that by calling `requireX` in the overriden `configure` function:

Reviewer 2: I think a sentence about the reason why `You should configure dependencies mostly in steps classes` would be beneficial for better understanding.

Author: better that way?

```
In order to have a plug-and-play-like experience you should configure dependencies mostly in steps classes (just in very rare cases you might also need to do it in test classes). You can add the dependency configuration by calling `requireX` in the overridden `configure` function:
```

Reviewer 2: Sounds good 👍

```kotlin
override fun configure() = super.configure()
    .requireReal<LoginViewModel>()
```

"require" here means that that you expect the type (in this case `LoginViewModel`) in its real form, not as a mock. So if the configuration says `mockOnly of<LoginViewModel>` or in another steps or test class you declare `requireMock<LoginViewModel>` there's a conflict and sweetest will throw an exception. Expectations about dependencies need to be unanbiguous in a test system.

> **Caution:** make sure you have declared the dependency type in the [module testing configuration](#adding-dependencies)!

Author: afaik in english it's lower-case, but in german it'd be correct 😁

> **Caution:** make sure you have declared the dependency type in the [module testing configuration](#adding-dependencies)!

Reviewer 3: it's more complicate than that.. also in english sometimes uppercase and here since its a noun I think.. but also not really important :D

##### Offering _and_ requiring at the same time

Offering is a different concept than requiring: that you offer the framework how to come to a certain dependency type doesn't mean it's configured to deliver it. E.g. you may `offerReal<AuthManager> { ... }` but forget to call `requireReal<AuthManager>()` somewhere in your test system. As a consequence you'll end up with a mock nevertheless.

But in order to avoid this problem consider using just the `offerXRequired` family of functions, as it caters for the most common use case:

```kotlin
offerRealRequired<AuthManager> { AuthManager(myBackendGateway, mySessionStore) }
offerMockRequired<AuthManager> { FakeAuthManager() }
```

The reason for being able to just "offer" something without actually putting it to use is that steps classes can also be seen as libraries which define how to create a real or mocked version of a type, where the decision what to pick finally can happen elsewhere. To keep things simple, in most cases `offerMockRequired` and `offerRealRequired` does the trick though.

##### Distinguish mock and real correctly

> Make sure to distinguish mock and real correctly: when you're using the production type as it's used in your product, consider it real. Everything else that is specific to your test system like mocks, spies and fakes should fall under the "mock" category for sweetest.

Reviewer 2: I first read it like _"how do I know if an given instance is a mock or real"_. But what it's actually about: _"When should I use real mode and when mock mode"_. Maybe this makes it a little clearer?:
```suggestion
Make sure to use mock and real modes correctly: when you're using the production type as it's used in your product, consider it real. Everything else that is specific to your test system like mocks, spies and fakes should fall under the "mock" category for sweetest.
```

Author: i moved it into the heading

#### Special case: abstract types and type hierarchies

When consuming or configuring dependencies sweetest tries to find a dependency declaration for the type in the module configurations:

```kotlin

> // in the module configuration:
> 
> dependency any of<BackendGateway>()
> 
> // e.g. in the steps class
> 
> private val instance by dependency<BackendGateway>() // <-- here
> 
> override fun configure() = super.configure()
>     .requireReal<BackendGateway>() // <-- here 

Reviewer 2: I'd make the `<-- here` more explicit, maybe:
```suggestion
// in the module configuration:

dependency any of<BackendGateway>() // <-- declared here

// e.g. in the steps class

private val instance by dependency<BackendGateway>() // <-- configured here

override fun configure() = super.configure()
    .requireReal<BackendGateway>() // <-- configured here 
```

Author: `by dependency` should be saying `used here` probably. now it's:

```
// in the module configuration:

dependency any of<BackendGateway>() // <-- declared here

// e.g. in the steps class

private val instance by dependency<BackendGateway>() // <-- used here

override fun configure() = super.configure()
    .requireReal<BackendGateway>() // <-- configured here 
```

    .offerMockRequired<BackendGateway> { instance }
```

This is easier but requires the class (in this case `FakeBackendGateway`) being compatible with it being initialized that early during creation of the steps classes. Feel free to use that approach when suitable.

#### Special case: spy

The approach described above is also of value if you need to create a spy on a class.

**Approach A:**

```kotlin
private val instance = spy(FakeBackendGateway())

override fun configure() = super.configure()
    .offerMockRequired<BackendGateway> { instance }
```

**Approach B:**

But it's also possible to create Mockito spies like so:

```kotlin
private val instance by dependency<AuthManager>()

override fun configure() = super.configure()
    .requireSpy<AuthManager>()
```

The downside using `requireSpy` is that you have no control over the creation of the underlying class that is spied on: an `AuthManager` will be created and wrapped by a spy.

In contrast to that, for the `FakeBackendGateway` we _need_ to use **approach A** as we need more control over the spy creation (we don't need a spy on `BackendGateway` but rather on `FakeBackendGateway`).

> ### Module testing configuration

Author: hm, i kept the mentions in the introduction on a surface level intentionally and linked the reference ("for more info see..."). where the intro is here to give a broad understanding, the reference is going into detail. your feedback makes me question whether that goal was really achieved now 🤔 can you probably explain a little more?

> ### Module testing configuration

Reviewer 3: ok I'll try again :)
Coming from the reading it top to bottom, as a whole, I felt like some information is repeating (same in intro and ref, maybe different words) and some is adding more depth in the reference but also some things are _better_ explained in the introduction.

As I understand you'd prefer to be the reference more stand-alone and the introduction is like an add-on.. 
I thought, thats why I also mentioned it before, as a whole document maybe a different structure would be beneficial (e.g. theory/basics, reference, tips, example/tutorial/intro) and to merge the spread information.

Another thing could be to just shorten the introduction to an example with one sentence to something like steps and always link only to the reference for more (there the information is merged on this topic), since the basic names and stuff have to be read anyways..

Hope I could make my thoughts/intentions clear.. it is really just something that came to mind and don't get me wrong it is just thinking out loud :)
